src/ejson2/EjsonEncoder.ts

```diff
@@ -0,0 +1,593 @@
+import {
+  BsonBinary,
+  BsonDbPointer,
+  BsonDecimal128,
+  BsonFloat,
+  BsonInt32,
+  BsonInt64,
+  BsonJavascriptCode,
+  BsonJavascriptCodeWithScope,
+  BsonMaxKey,
+  BsonMinKey,
+  BsonObjectId,
+  BsonSymbol,
+  BsonTimestamp,
+} from '../bson/values';
+import {toBase64Bin} from '@jsonjoy.com/base64/lib/toBase64Bin';
+import {Writer} from '@jsonjoy.com/util/lib/buffers/Writer';
+import type {IWriter, IWriterGrowable} from '@jsonjoy.com/util/lib/buffers';
+import type {BinaryJsonEncoder} from '../types';
+
+export interface EjsonEncoderOptions {
+  /** Use canonical format (preserves all type information) or relaxed format (more readable) */
+  canonical?: boolean;
+}
+
+export class EjsonEncoder implements BinaryJsonEncoder {
```

Comment:
Make this class extends `JsonEncoder`. Reuse parent class methods, where it makes sense.